src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+
+        internal Stack<DisposeScope> DisposeScopeStack { get; } = new();
+        internal IDisposable CurrentlyDisposing { get; set; }
+
+        public int CreatedCount => _createdCount;
+        public int DisposedCount => _disposedCount;
+        public int TotalCount => _createdCount - _disposedCount;
+
+        public static long TotalAllocatedSize => Singleton.InnerTotalAllocatedSize;
+
+        internal static void TryUnregisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryUnregisterInDisposeScope(disposable);
+
+        internal static void TryRegisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryRegisterInDisposeScope(disposable);
+
+        private long InnerTotalAllocatedSize {
+            get {
+                long size = 0;
+                lock (this) {
+                    foreach (var disposeScope in DisposeScopeStack) {
+                        size += disposeScope.TotalAllocatedSize;
+                    }
+                }
+
+                return size;
+            }
+        }
+
+        private void InnerTryRegisterInDisposeScope(IDisposable disposable)
+        {
+            Interlocked.Increment(ref _createdCount);
+
+            lock (this) {
+                if (DisposeScopeStack.Count > 0) {
+                    DisposeScopeStack.Peek().Include(disposable);
+                }
+            }
+        }
+
+        private void InnerTryUnregisterInDisposeScope(IDisposable disposable)
+        {
+            // It's us! We're disposing of it.
+            if (disposable == CurrentlyDisposing) {
+                return;
+            }
+
+            Interlocked.Increment(ref _disposedCount);
+            lock (this) {
+                foreach (var disposeScope in DisposeScopeStack) {
+                    disposeScope.Disposables.Remove(disposable);
+                }
+            }
+        }
+
+        internal static DisposeScope NewDisposeScope()
+        {
+            lock (Singleton) {
+                return Singleton.InnerNewDisposeScope();
+            }
+        }
+
+        private DisposeScope InnerNewDisposeScope()
+        {
+            lock (this) {
+                var disposeScope = new DisposeScope(this);
+                DisposeScopeStack.Push(disposeScope);
+                return disposeScope;
+            }
+        }
+
+        internal void RemoveDisposeScope(DisposeScope disposeScope)
+        {
+            lock (this) {
+                Debug.Assert(DisposeScopeStack.Count > 0);
+                Debug.Assert(DisposeScopeStack.Peek() == disposeScope);
+                DisposeScopeStack.Pop();
+            }
+        }
+
+        internal void MoveToOuterScope(DisposeScope disposeScope, IDisposable disposable)
+        {
+            lock (this) {
+                var array = DisposeScopeStack.ToArray();
+                for (var i = array.Length - 1; i >= 1; i--) {
+                    if (array[i] == disposeScope) {
```

Comment:
Aren't you keeping parent scope in `DisposeScope`? If not, it might be a good idea to do so.

Reply:
I am now! It was a great idea